flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java

```diff
@@ -386,6 +387,19 @@ private void checkPatternNameUniqueness(final Pattern pattern) {
 			return copyOfSink;
 		}
 
+		private State<T> copy(final State<T> state) {
+			final State<T> copyOfState = new State<>(state.getName(), state.getStateType());
```

Comment:
The copy should be made with `createState` method which does ensure there is a unique name assigned, which is neccessary for the serialization.